File: read_dicomct.py
"""
read_dicomct - Read DICOM CT and apply IEC convention

Reads DICOM CT images from low to high Y and fills the structure with images 
according to the IEC convention.

This CT is given in the image coordinate system:
              -----------         IEC
             /|         /|         Z
            / |        / |        /|
           /  |       /  |         |   /| Y
          /   |      /   |         |   /
         /    ------/----          |  /
         ----/------    /          | /
        |   /      |   /           |/
        |  /       |  /     X------|----------------->
        | /        | /            /|
        |/         |/            / |
        ------------

PixelFirst coordinates will be the bottom-left-corner. The CT cube will
be addressed in the following way CT.Image(1:Width,1:Depth,1:Height) or
CT.Image(X,Y,Z)

History
-------
Original MATLAB version by Andre Dekker, Lucas Persoon, Wouter van Elmpt, 
Bas Nijsten @ MAASTRO (2005-2012)
Converted to Python
"""

import numpy as np
import pydicom


def read_dicomct(filenames_in, read_image_data=True):
    """
    Read DICOM CT and apply IEC convention.
    
    Parameters
    ----------
    filenames_in : list of str
        Full filenames of the CT DICOM files
    read_image_data : bool, optional
        If True, read the actual CT image data. Default is True.
        
    Returns
    -------
    dict or None
        Dictionary containing CT information:
        - Filenames : list of str - Sorted filenames
        - UIDs : list of str - SOP instance UIDs
        - DicomHeader : pydicom.Dataset - DICOM header of first slice
        - PixelSpacingXi : float - Pixel spacing in cm/pixel X
        - PixelSpacingYi : float - Pixel spacing in cm/pixel Y  
        - PixelSpacingZi : float - Pixel spacing in cm/pixel Z
        - PixelNumXi : int - Number of pixels X
        - PixelNumYi : int - Number of pixels Y
        - PixelNumZi : int - Number of pixels Z
        - PixelFirstXi : float - X value of pixel 1 in cm
        - PixelFirstYi : float - Y value of pixel 1 in cm
        - PixelFirstZi : float - Z value of pixel 1 in cm
        - RescaleIntercept : float - Offset to convert to Hounsfield units
        - RescaleSlope : float - Slope to convert to Hounsfield units
        - Model : str - Manufacturer model name
        - Manufacturer : str - Manufacturer
        - Institute : str - Institution name
        - TableHeight : float - Table height
        - PrivTableHeight : float - Private table height
        - MachineName : str - Machine name (if available)
        - HUToRED : ndarray - HU to RED conversion (if available)
        - Image : ndarray - CT image (if read_image_data=True)
        
        Returns None if invalid input is provided.
    """
    slice_num = len(filenames_in)
    
    # Check for valid input
    if slice_num == 2 and '.' in filenames_in[0] and '..' in filenames_in[1]:
        return None
    
    # Sort the DICOM files from low to high slice Y value
    yi = []
    uids = []
    
    for slice_cur in range(slice_num):
        dicom_header = pydicom.dcmread(filenames_in[slice_cur], stop_before_pixels=True)
        yi.append([slice_cur, float(dicom_header.ImagePositionPatient[2])])
        uids.append(dicom_header.SOPInstanceUID)
    
    # Get unique Y positions
    yi = np.array(yi)
    _, unique_indices = np.unique(yi[:, 1], return_index=True)
    yi = yi[unique_indices]
    slice_num = len(yi)
    
    # Sort by Y position
    sorted_indices = np.argsort(yi[:, 1])
    slice_value_y_sorted = yi[sorted_indices]
    
    # Create output structure
    ctout = {}
    ctout['Filenames'] = [filenames_in[int(idx)] for idx in slice_value_y_sorted[:, 0]]
    ctout['UIDs'] = [uids[int(idx)] for idx in slice_value_y_sorted[:, 0]]
    
    # Read the first DICOM header
    ctout['DicomHeader'] = pydicom.dcmread(ctout['Filenames'][0], stop_before_pixels=True)
    
    # Store pixel spacing information (convert mm to cm)
    ctout['PixelSpacingXi'] = float(ctout['DicomHeader'].PixelSpacing[0]) / 10.0
    
    # Calculate Y spacing from slice positions
    if len(yi) > 1:
        y_slice_thicknesses = np.abs(np.unique(np.round((yi[:-1, 1] - yi[1:, 1]) * 1000) / 1000))
        if len(y_slice_thicknesses) > 0:
            ctout['PixelSpacingYi'] = y_slice_thicknesses[0] / 10.0
        else:
            ctout['PixelSpacingYi'] = float(ctout['DicomHeader'].SliceThickness) / 10.0
    else:
        ctout['PixelSpacingYi'] = float(ctout['DicomHeader'].SliceThickness) / 10.0
    
    ctout['PixelSpacingZi'] = float(ctout['DicomHeader'].PixelSpacing[1]) / 10.0
    
    # Store image dimensions
    ctout['PixelNumXi'] = int(ctout['DicomHeader'].Columns)
    ctout['PixelNumYi'] = len(ctout['Filenames'])
    ctout['PixelNumZi'] = int(ctout['DicomHeader'].Rows)
    
    # Calculate first pixel positions using IEC convention
    image_orientation = np.array(ctout['DicomHeader'].ImageOrientationPatient)
    reference_orientation = np.array([1, 0, 0, 0, 1, 0])
    
    if np.sum(np.abs(image_orientation - reference_orientation) > 0.025) == 0:
        image_position = np.array(ctout['DicomHeader'].ImagePositionPatient)
        
        ctout['PixelFirstXi'] = (image_position[0] / 10.0) - \
                                (image_orientation[0] == -1) * \
                                (ctout['PixelSpacingXi'] * ctout['PixelNumXi'])
        
        ctout['PixelFirstYi'] = image_position[2] / 10.0
        
        ctout['PixelFirstZi'] = (-image_position[1] / 10.0) - \
                                image_orientation[4] * \
                                (ctout['PixelSpacingZi'] * (ctout['PixelNumZi'] - 1))
    else:
        ctout['PixelFirstXi'] = None
        ctout['PixelFirstYi'] = None
        ctout['PixelFirstZi'] = None
    
    # Store rescale parameters
    ctout['RescaleIntercept'] = float(ctout['DicomHeader'].RescaleIntercept)
    ctout['RescaleSlope'] = float(ctout['DicomHeader'].RescaleSlope)
    
    # Store manufacturer information
    if hasattr(ctout['DicomHeader'], 'ManufacturerModelName'):
        ctout['Model'] = ctout['DicomHeader'].ManufacturerModelName
    else:
        ctout['Model'] = None
    
    ctout['Manufacturer'] = ctout['DicomHeader'].Manufacturer
    
    # Store institution information
    ctout['Institute'] = ''
    if hasattr(ctout['DicomHeader'], 'InstitutionName'):
        if 'Maastro' in ctout['DicomHeader'].InstitutionName:
            ctout['Institute'] = 'Maastro'
    
    # Store table height
    if hasattr(ctout['DicomHeader'], 'TableHeight'):
        ctout['TableHeight'] = float(ctout['DicomHeader'].TableHeight)
    else:
        ctout['TableHeight'] = 0.0
    
    # Store private table height from various private tags
    ctout['PrivTableHeight'] = 0.0
    
    private_tags = [
        ('Private_1099_10xx_Creator', 'Private_1099_1099'),
        ('Private_1199_11xx_Creator', 'Private_1199_1199'),
        ('Private_1299_12xx_Creator', 'Private_1299_1299'),
        ('Private_1399_13xx_Creator', 'Private_1399_1399'),
    ]
    
    for creator_tag, data_tag in private_tags:
        if hasattr(ctout['DicomHeader'], creator_tag) and hasattr(ctout['DicomHeader'], data_tag):
            value = getattr(ctout['DicomHeader'], data_tag)
            if isinstance(value, str):
                ctout['PrivTableHeight'] = float(value)
            else:
                ctout['PrivTableHeight'] = float(str(value))
            break
    
    # Store machine name
    if hasattr(ctout['DicomHeader'], 'Private_1499_14xx_Creator') and \
       hasattr(ctout['DicomHeader'], 'Private_1499_1499'):
        value = ctout['DicomHeader'].Private_1499_1499
        if isinstance(value, str):
            ctout['MachineName'] = value.strip()
        else:
            ctout['MachineName'] = str(value).strip()
    else:
        ctout['MachineName'] = None
    
    # Store HU to RED conversion if available
    if hasattr(ctout['DicomHeader'], 'Private_1599_15xx_Creator') and \
       hasattr(ctout['DicomHeader'], 'Private_1599_1599'):
        try:
            hu_to_red_string = ctout['DicomHeader'].Private_1599_1599
            if isinstance(hu_to_red_string, bytes):
                # Convert bytes to array of doubles
                hu_to_red = []
                for i in range(0, len(hu_to_red_string), 8):
                    if i + 8 <= len(hu_to_red_string):
                        value = np.frombuffer(hu_to_red_string[i:i+8], dtype=np.float64)[0]
                        hu_to_red.append(value)
                ctout['HUToRED'] = np.array(hu_to_red)
            else:
                ctout['HUToRED'] = None
        except:
            ctout['HUToRED'] = None
    else:
        ctout['HUToRED'] = None
    
    # Read all slices if requested
    if read_image_data:
        # Initialize image array (Z, X, Y) initially
        ctout['Image'] = np.zeros((ctout['PixelNumZi'], 
                                   ctout['PixelNumXi'], 
                                   ctout['PixelNumYi']))
        
        # Read and apply RescaleSlope and RescaleIntercept
        for slice_cur in range(slice_num):
            pixel_array = pydicom.dcmread(ctout['Filenames'][slice_cur]).pixel_array
            ctout['Image'][:, :, slice_cur] = (pixel_array.astype(float) * 
                                               ctout['RescaleSlope'] + 
                                               ctout['RescaleIntercept'])
        
        # Clip values below -1024 to -1024
        ctout['Image'][ctout['Image'] < -1024] = -1024
        
        # Apply IEC format transformations
        # Flip Z axis (reverse first dimension)
        ctout['Image'] = ctout['Image'][::-1, :, :]
        
        # Permute to (X, Y, Z) format
        # From (Z, X, Y) to (X, Z, Y)
        ctout['Image'] = np.transpose(ctout['Image'], (1, 0, 2))
        # From (X, Z, Y) to (X, Y, Z)
        ctout['Image'] = np.transpose(ctout['Image'], (0, 2, 1))
    
    return ctout

File: read_dicomct.m
function CTOUT   =   read_dicomct(FilenamesIN,varargin)
% function CTOUT   =   read_dicomct(FilenamesIN)
% Reads a CT and applies IEC convention
% This function reads the DICOM images from low to high Y and fills the
% structure CTOUT with the images according to the IEC convention.

% This CT is given in the image coordinate system
%               -----------         IEC
%              /|         /|         Z
%             / |        / |        /|\
%            /  |       /  |         |   /|\ Y
%           /   |      /   |         |   /
%          /    ------/----          |  /
%          ----/------    /          | /
%         |   /      |   /           |/
%         |  /       |  /     X------|----------------->
%         | /        | /            /|
%         |/         |/            / |
%         ------------
%
% PixelFirst coordinates will be the bottom-left-corner. The CT cube will
% be addressed in the following way CT.Image(1:Width,1:Depth,1:Height) Or
% CT.Image(X,Y,Z)
%--------------------------------------------------------------------------
% INPUT
%--------------------------------------------------------------------------
% NAME              TYPE        DESCRIPTION 
% FilenamesIN       cell        Full filenames of the CT dicom files
%--------------------------------------------------------------------------
% OUTPUT
%--------------------------------------------------------------------------
% NAME              TYPE        DESCRIPTION 
% CTOUT             struct      
% .Filenames        cell        Filenames of the dicom files
% .DicomHeader      struct      The dicomheader of the first slice
% .PixelSpacingXi   double      Pixel spacing in cm/pixel X in the i cs
% .PixelSpacingYi   double      Pixel spacing in cm/pixel Y in the i cs
% .PixelSpacingZi   double      Pixel spacing in cm/pixel Z in the i cs
% .PixelNumXi       double      Number of pixels X in the i cs
% .PixelNumYi       double      Number of pixels Y in the i cs
% .PixelNumZi       double      Number of pixels Z in the i cs
% .PixelFirstXi     double      X value of pixel 1 in cm in the i cs
% .PixelFirstYi     double      Y value of pixel 1 in cm the i cs
% .PixelFirstZi     double      Z value of pixel 1 in cm the i cs
% .RescaleIntercept double      Offset to convert to hounsfield units
% .RescaleSlope     double      Slope to convert to hounsfield units
% .Image            uint16      CT Image
%--------------------------------------------------------------------------
% INTERNAL VARIABLE DEFINITION
%--------------------------------------------------------------------------
% NAME              TYPE        DESCRIPTION 
% DicomFiles        struct      Dicom file info
% Filenames         cell        Dicom filenames
% SliceNum          double      Number of slices
% SliceCur          double      Currrent slice
% DicomHeader       struct      Dicom header
% Yi                double      Value of slice in the i cs
% SliceValueYSorted double      Slices sorted from low Y to high Y
% CouchTopCorr      double      Thickness of the couch top in cm
% TopTableIndex     double      Pixel indicating the top of the table
%
%--------------------------------------------------------------------------
% FUNCTIONS & MEX files called
%--------------------------------------------------------------------------
% none
%--------------------------------------------------------------------------
% HISTORY
%--------------------------------------------------------------------------
% 08/09/2005, Andre Dekker
% Creation
%
% 13/09/2005, Lucas Persoon & Andre Dekker
% Added zero pixel intensities beneath couch top part, this deletes the
% couch from the picture. At a later time we have to check if this
% correction is sufficient.
% 
% 15/09/2005, Andre Dekker
% Changed variable names and comment to comply to standard.
%
% 19/09/2005, Andre Dekker
% Changed of to switch statement.
%
% 20/09/2005, Andre Dekker
% Removed table pixels from CT to speed up program
%
% 28/09/2005, Andre Dekker
% Redefined PixelFirstXi and PixelFirstZi 
%
% 30/09/2005, Andre Dekker
% Removed couch pixel deletion, made a seperate function
%
% 21/10/2005, Andre Dekker
% Modified call of read_dicomct, form path to a cell with the filenames
%
% 23/12/2005, Lucas Persoon
% Added a fail safe when no CT is in the directory it does nothing instead 
% of erroring
% 
% 05/01/2006, Lucas Persoon
% Changed the column width to 80
% 
% 08/02/2006, Andre Dekker
% Corrected SAMPLE CALL
%
% 01/08/2006, Lucas Persoon
% Added comment by the rotation of the matrix
% rotate matrix around x axis -90 ccw
% is necessary for IEC convention means a flip in de slice and
% reverse of z axis
% see page 137-139 IEC 61217
%
% 15/08/2006, Lucas Persoon 
% Made the PixelFirst IEC conforming and confming the new XiO version
%
% 25/09/2006, Lucas Persoon
% Changed the transformations according to the IEC-DICOM manual of MAASTRO
% created by Bas Nijsten and Lucas Persoon
%
% 18/10/2006, Lucas Persoon 
% Addded the manufacturer, institutename, tableheigth en private tableheight
% to the struct
%
% 01/11/2006, Wouter van Elmpt
% Changed the PixelFirstZi with the correct function -> (PixelNumZi-1)*PixelSpacingZ
%
% 08/12/2010, Bas Nijsten
% Added the SOP instance UIDs to the struct
%
% 27/09/2012, Bas Nijsten
% Changed the ImageOrientationPatient tag handling
%--------------------------------------------------------------------------
% SAMPLE CALL
%--------------------------------------------------------------------------
% PathCT                      =   uigetdir('C:\',...
%                                 'Select directory with CT dicom files');
% DicomFiles                  =   dir([PathCT,'\*.dcm']);
% for i=1:length(DicomFiles)
%     Filenames{i}            =   fullfile(PathCT,DicomFiles(i).name);
% end
% CT                          =   read_dicomct(Filenames);
%--------------------------------------------------------------------------

SliceNum                 =   uint16(length(FilenamesIN));
if ~(SliceNum == 2 && ~isempty(strfind(FilenamesIN{1},'\.')) && ...
    ~isempty(strfind(FilenamesIN{2},'\..')))
%beginif
%read CT otherwise something is wrong in the filenames and nothing is done

% Sort the DICOM files from low to high slice Y value, and store filenames
% and SOP instance UIDs in structure
for SliceCur=1:SliceNum
    DicomHeader          =   dicominfo(FilenamesIN{SliceCur});
    Yi(SliceCur,:)       =   [double(SliceCur),...
                             DicomHeader.ImagePositionPatient(3)];
    UIDs{SliceCur,:}     =   DicomHeader.SOPInstanceUID;
end %for SliceCur
[~,I]                    =   unique(Yi(:,2));
Yi                       =   Yi(I,:);
SliceNum                 =   size(Yi,1);
SliceValueYSorted        =   sortrows(Yi,2);
CTOUT.Filenames          =   FilenamesIN(SliceValueYSorted(:,1));
CTOUT.UIDs               =   UIDs(SliceValueYSorted(:,1))';

% Read the first Dicom header and store in structure
CTOUT.DicomHeader        =   dicominfo(CTOUT.Filenames{1});

% Store information from the Dicom header in structure for later quick
% access
CTOUT.PixelSpacingXi     =   CTOUT.DicomHeader.PixelSpacing(1)/10; 
                             %/10 for mm to cm (IEC)
YSliceThicknesses        =   abs(unique(round((Yi(1:end-1,2)-Yi(2:end,2))*1000)/1000));
if ~ isempty(YSliceThicknesses)
    CTOUT.PixelSpacingYi =   YSliceThicknesses(1)/10;
else
    CTOUT.PixelSpacingYi =   CTOUT.DicomHeader.SliceThickness/10;
end 
                             %/10 for mm to cm (IEC)
CTOUT.PixelSpacingZi     =   CTOUT.DicomHeader.PixelSpacing(2)/10; 
                             %/10 for mm to cm (IEC)
CTOUT.PixelNumXi         =   double(CTOUT.DicomHeader.Width);
CTOUT.PixelNumYi         =   double(length(CTOUT.Filenames));
CTOUT.PixelNumZi         =   double(CTOUT.DicomHeader.Height);
% if CTOUT.DicomHeader.ImageOrientationPatient == [ -1 ; 0 ; 0 ; 0 ; -1 ; 0 ] 
%     CTOUT.PixelFirstXi     =(CTOUT.DicomHeader.ImagePositionPatient(1)/10)-...
%                             (CTOUT.PixelSpacingXi*CTOUT.PixelNumXi); 
%                              %/10 for mm to cm (IEC)
%     CTOUT.PixelFirstYi     =(CTOUT.DicomHeader.ImagePositionPatient(3)/10);
%                                 %/10 for mm to cm (IEC)
%     CTOUT.PixelFirstZi     =-((CTOUT.DicomHeader.ImagePositionPatient(2)/10) -...
%                             (CTOUT.PixelSpacingZi*(CTOUT.PixelNumZi-1)));    
% else 
%     CTOUT.PixelFirstXi     =CTOUT.DicomHeader.ImagePositionPatient(1)/10; 
%                                 %/10 for mm to cm (IEC)
%     CTOUT.PixelFirstYi     =CTOUT.DicomHeader.ImagePositionPatient(3)/10; 
%                                 %/10 for mm to cm (IEC)
%     CTOUT.PixelFirstZi     =(-CTOUT.DicomHeader.ImagePositionPatient(2)/10) -...
%                             (CTOUT.PixelSpacingZi*(CTOUT.PixelNumZi-1));
% end
if sum(abs(CTOUT.DicomHeader.ImageOrientationPatient - [ 1 ; 0 ; 0 ; 0 ; 1 ; 0 ]) > 0.025) == 0
    CTOUT.PixelFirstXi   =   (CTOUT.DicomHeader.ImagePositionPatient(1)/10)-...
                                (CTOUT.DicomHeader.ImageOrientationPatient(1) == -1)*...
                                (CTOUT.PixelSpacingXi*CTOUT.PixelNumXi); 
                                %/10 for mm to cm (IEC)
    CTOUT.PixelFirstYi   =   (CTOUT.DicomHeader.ImagePositionPatient(3)/10);
                                %/10 for mm to cm (IEC)
    CTOUT.PixelFirstZi   =   (-CTOUT.DicomHeader.ImagePositionPatient(2)/10)-...
                                CTOUT.DicomHeader.ImageOrientationPatient(5)*...
                                (CTOUT.PixelSpacingZi*(CTOUT.PixelNumZi-1)); 
                                %/10 for mm to cm (IEC)
else
    CTOUT.PixelFirstXi   =   [];
    CTOUT.PixelFirstYi   =   [];
    CTOUT.PixelFirstZi   =   [];
end
    
CTOUT.RescaleIntercept   =   CTOUT.DicomHeader.RescaleIntercept;
CTOUT.RescaleSlope       =   CTOUT.DicomHeader.RescaleSlope;
if isfield(CTOUT.DicomHeader,'ManufacturerModelName')
    CTOUT.Model          =   CTOUT.DicomHeader.ManufacturerModelName;
else
    CTOUT.Model          =   [];
end
CTOUT.Manufacturer       =   CTOUT.DicomHeader.Manufacturer;

CTOUT.Institute = '';
if isfield(CTOUT.DicomHeader,'InstitutionName');
    if strfind(CTOUT.DicomHeader.InstitutionName,'Maastro')
        CTOUT.Institute = 'Maastro';
    end
end

if isfield(CTOUT.DicomHeader,'TableHeight');
   CTOUT.TableHeight     =   CTOUT.DicomHeader.TableHeight;
else
   CTOUT.TableHeight     =   0;
end

CTOUT.PrivTableHeight    =   0;

if isfield(CTOUT.DicomHeader,'Private_1099_10xx_Creator'),
    if ischar(CTOUT.DicomHeader.Private_1099_1099)
        CTOUT.PrivTableHeight=str2num(CTOUT.DicomHeader.Private_1099_1099);
    else
        CTOUT.PrivTableHeight=str2num(char(CTOUT.DicomHeader.Private_1099_1099)');
    end
end
if isfield(CTOUT.DicomHeader,'Private_1199_11xx_Creator'),
    if ischar(CTOUT.DicomHeader.Private_1199_1199)
        CTOUT.PrivTableHeight=str2num(CTOUT.DicomHeader.Private_1199_1199);
    else
        CTOUT.PrivTableHeight=str2num(char(CTOUT.DicomHeader.Private_1199_1199)');
    end
end
if isfield(CTOUT.DicomHeader,'Private_1299_12xx_Creator'),
    if ischar(CTOUT.DicomHeader.Private_1299_1299)
        CTOUT.PrivTableHeight=str2num(CTOUT.DicomHeader.Private_1299_1299);
    else
        CTOUT.PrivTableHeight=str2num(char(CTOUT.DicomHeader.Private_1299_1299)');
    end
end
if isfield(CTOUT.DicomHeader,'Private_1399_13xx_Creator'),
    if ischar(CTOUT.DicomHeader.Private_1399_1399)
        CTOUT.PrivTableHeight=str2num(CTOUT.DicomHeader.Private_1399_1399);
    else
        CTOUT.PrivTableHeight=str2num(char(CTOUT.DicomHeader.Private_1399_1399)');
    end
end

if isfield(CTOUT.DicomHeader,'Private_1499_14xx_Creator'),
    if ischar(CTOUT.DicomHeader.Private_1499_1499)
        CTOUT.MachineName=deblank(CTOUT.DicomHeader.Private_1499_1499);
    else
        CTOUT.MachineName=deblank(char(CTOUT.DicomHeader.Private_1499_1499)');
    end
end 

if isfield(CTOUT.DicomHeader,'Private_1599_15xx_Creator') && strcmp(class(CTOUT.DicomHeader.Private_1599_1599),'uint8')
    HUToREDString = CTOUT.DicomHeader.Private_1599_1599';
    idxVect  = 1:8:length(HUToREDString);
    for i = 1:length(idxVect)
        SF         = typecast(uint8(HUToREDString(idxVect(i):idxVect(i)+7)), lower('double'));
        HUToRED(i) = SF;
    end
    CTOUT.HUToRED  = HUToRED;
else
    CTOUT.HUToRED = [];
end
    
% Read all slices in path, convert to Hounsfield 
% Units, apply IEC convention
if ~ (~ isempty(varargin) && ~ isempty(varargin{1}) && ~ isempty(varargin{2}) && ...
    strcmp(varargin{1},'readimagedata') && varargin{2}==false)
    CTOUT.Image             =   zeros(CTOUT.PixelNumZi,...
        CTOUT.PixelNumXi,CTOUT.PixelNumYi);
    
    % Apply RescaleSlope and RescaleIntercept
    for SliceCur = 1 : SliceNum
        CTOUT.Image(:,:,SliceCur) = ...
            double(dicomread(CTOUT.Filenames{SliceCur}))*CTOUT.RescaleSlope + ...
            CTOUT.RescaleIntercept;
    end
    
    % Change all pixel values smaller than -1024 to -1024
    CTOUT.Image(find(CTOUT.Image<-1024)) = -1024;
    
    %Set CT in IEC format
    CTOUT.Image(:,:,:)            =   CTOUT.Image(end:-1:1,:,:);
    CTOUT.Image                   =   permute(CTOUT.Image,[ 2 1 3 ]);
    CTOUT.Image                   =   permute(CTOUT.Image,[ 1 3 2 ]);
end
end%endif

